nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java

```diff
@@ -0,0 +1,322 @@
+  private static final Logger logger = LoggerFactory.getLogger(NessieCatalog.class);
+  private static final Joiner SLASH = Joiner.on("/");
+  public static final String NESSIE_WAREHOUSE_DIR = "nessie.warehouse.dir";
+  private NessieClient client;
+  private String warehouseLocation;
+  private Configuration config;
+  private UpdateableReference reference;
+  private String name;
+  private FileIO fileIO;
+
+  public NessieCatalog() {
+  }
+
+  @Override
+  public void initialize(String inputName, Map<String, String> options) {
+    String fileIOImpl = options.get(CatalogProperties.FILE_IO_IMPL);
+    this.fileIO = fileIOImpl == null ? new HadoopFileIO(config) : CatalogUtil.loadFileIO(fileIOImpl, options, config);
+    this.name = inputName == null ? "nessie" : inputName;
+    this.client = NessieClient.withConfig(options::get);
+
+    this.warehouseLocation = options.get(NESSIE_WAREHOUSE_DIR);
+    if (warehouseLocation == null) {
+      throw new IllegalStateException("Parameter nessie.warehouse.dir not set, nessie can't store data.");
+    }
+    final String requestedRef = options.get(NessieClient.CONF_NESSIE_REF);
+    this.reference = loadReference(requestedRef);
+  }
+
+  @Override
+  public void close() {
+    client.close();
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    TableReference pti = TableReference.parse(tableIdentifier);
+    UpdateableReference newReference = this.reference;
+    if (pti.getReference() != null) {
+      newReference = loadReference(pti.getReference());
+    }
+    return new NessieTableOperations(NessieUtil.toKey(pti.getTableIdentifier()), newReference, client, fileIO);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier table) {
+    if (table.hasNamespace()) {
+      return SLASH.join(warehouseLocation, table.namespace().toString(), table.name());
+    }
+    return SLASH.join(warehouseLocation, table.name());
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    return tableStream(namespace).collect(Collectors.toList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    reference.checkMutable();
+
+    IcebergTable existingTable = table(identifier);
+    if (existingTable == null) {
+      return false;
+    }
+
+    // We try to drop the table. Simple retry after ref update.
+    int count = 0;
+    while (count < 5) {
+      count++;
+      try {
+        dropTableInner(identifier);
+        break;
+      } catch (NessieConflictException e) {
+        // pass for retry
+      } catch (NessieNotFoundException e) {
+        logger.error("Cannot drop table: ref is no longer valid.", e);
+        return false;
+      }
+    }
+    if (count >= 5) {
+      logger.error("Cannot drop table: failed after retry (update hash and retry)");
+      return false;
+    }
```

Comment:
You can replace this with `Tasks`:

```java
Tasks.foreach(identifier)
    .retry(5)
    .stopRetryOn(NessieNotFoundException.class)
    .throwFailureWhenFinished()
    .run(this::dropTableInner)
```

`Tasks` is pretty flexible and allows you to configure exponential backoff, whether to retry on all exceptions but a known list (`stopRetryOn`), whether to retry on specific exceptions (`onlyRetryOn`), and set up callbacks (`onFailure`).

It helps us avoid logic like this everywhere.

Reply:
:+1: